x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,82 +351,133 @@ public void removeFromCache(final CacheKey cacheKey) {
+     * Evicts the cache files associated to the specified {@link ShardEviction}.
      *
-     * @param shardEviction a {@link ShardEviction} representing the shard marked as evicted
-     * @param runnable      a runnable to execute
+     * @param shardEviction the shard eviction to process
      */
-    private void runIfShardMarkedAsEvictedInCache(ShardEviction shardEviction, Runnable runnable) {
-        try (Releasable ignored = shardsEvictionLock.acquire(shardEviction)) {
-            boolean success = false;
+    void processShardEviction(ShardEviction shardEviction) {
+        assert isPendingShardEviction(shardEviction) : "shard is not marked as evicted: " + shardEviction;
+        assert assertGenericThreadPool();
+
+        shardsEvictionsLock.readLock().lock();
+        try {
             try {
-                if (evictedShards.remove(shardEviction)) {
-                    runnable.run();
+                if (allowShardsEvictions) {
+                    final List<CacheFile> cacheFilesToEvict = new ArrayList<>();
+                    cache.forEach((cacheKey, cacheFile) -> {
+                        if (shardEviction.matches(cacheKey)) {
+                            cacheFilesToEvict.add(cacheFile);
+                        }
+                    });
+                    for (CacheFile cacheFile : cacheFilesToEvict) {
+                        try {
+                            cache.invalidate(cacheFile.getCacheKey(), cacheFile);
+                        } catch (RuntimeException e) {
+                            logger.warn(() -> new ParameterizedMessage("failed to evict cache file {}", cacheFile.getCacheKey()), e);
+                            assert false : e;
+                        }
+                    }
+                    logger.debug(
+                        "shard eviction [{}] processed with [{}] cache files invalidated",
+                        shardEviction,
+                        cacheFilesToEvict.size()
+                    );
                 }
-                success = true;
             } finally {
-                if (success == false) {
-                    final boolean added = evictedShards.add(shardEviction);
-                    assert added : shardEviction;
+                synchronized (shardsEvictionsMutex) {
+                    final Future<?> removedFuture = pendingShardsEvictions.remove(shardEviction);
+                    assert removedFuture != null;
                 }
             }
+        } finally {
+            shardsEvictionsLock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Processes and waits for all pending shard evictions to complete.
+     */
+    private void processAllPendingShardsEvictions() {
+        final List<Future<?>> pendingFutures;
+        synchronized (shardsEvictionsMutex) {
+            allowShardsEvictions = false;
+            pendingFutures = new ArrayList<>(pendingShardsEvictions.values());
+        }
+        if (pendingFutures.isEmpty() == false) {
```

Comment:
I think this check is unnecessary now, if there are no pending shard evictions, grabbing the write lock should be really quick anyway.

Reply:
:+1: I pushed 6c29e73de133eed7e7da21e8c9aea548c578f479